networkoperator/v1/types_egressrouter.go

```diff
@@ -0,0 +1,274 @@
+package v1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// EgressRouter is a feature allowing the user to define an egress router
+// that acts as a bridge between pods and external systems. The egress router runs
+// a service that redirects egress traffic originating from a pod or a group of
+// pods to a remote external system or multiple destinations as per configuration.
+//
+// It is consumed by the cluster-network-operator.
+// More specifically, given an EgressRouter CR with <name>, the CNO will create and manage:
+// - A service called <name>
+// - An egress pod called <name>
+// - A NAD called <name>
+//
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+//
+// EgressRouter is a single egressrouter pod configuration object.
+// +k8s:openapi-gen=true
+// +kubebuilder:resource:path=egressrouters,scope=Namespaced
+// +kubebuilder:printcolumn:name="Condition",type=string,JSONPath=".status.conditions[*].type"
+// +kubebuilder:printcolumn:name="Status",type=string,JSONPath=".status.conditions[*].status"
+type EgressRouter struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	// Specification of the desired egress router.
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MinProperties:=4
+	// +kubebuilder:validation:MaxProperties:=4
+	Spec EgressRouterSpec `json:"spec" protobuf:"bytes,2,opt,name=spec"`
+
+	// Observed status of EgressRouter.
+	// +kubebuilder:validation:Optional
+	// +optional
+	Status EgressRouterStatus `json:"status,omitempty" protobuf:"bytes,3,opt,name=status"`
+}
+
+// EgressRouterSpec contains the configuration for an egress router.
+// Mode, networkInterface and addresses fields must be specified along with exactly one "Config" that matches the mode.
+// Each config consists of parameters specific to that mode.
+// +k8s:openapi-gen=true
+// +kubebuilder:validation:Required
+type EgressRouterSpec struct {
+	// Mode depicts the mode that is used for the egress router. The default mode is "Redirect" and is the only supported mode currently.
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Enum="Redirect"
+	// +kubebuilder:default:="Redirect"
+	Mode EgressRouterMode `json:"mode" protobuf:"bytes,1,opt,name=mode,casttype=EgressRouterMode"`
+
+	// Redirect represents the configuration parameters specific to redirect mode.
+	// +kubebuilder:validation:Optional
+	Redirect *RedirectConfig `json:"redirect,omitempty" protobuf:"bytes,2,opt,name=redirect"`
+
+	// Specification of interface to create/use. The default is macvlan.
+	// Currently only macvlan is supported.
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MinProperties:=1
+	// +kubebuilder:validation:MaxProperties:=1
```

Comment:
drop the min/maxProperties. What should they mean?

Reply:
ditto, in future we maybe add other interface types and then only one of such config's should be provided at a time in the yaml.